### PR TITLE
fix: AutoSuggestBox.QuerySubmitted ChosenSuggestion should not be null, patch.

### LINKED ISSUE #7778: [WASM][WindowsAppSDK] `AutoSuggestBox.QuerySubmitted` `ChosenSuggestion` should not be null
### Current behavior

Please see attached sample - which is a modified version of what I posted at #4382 - and search for '!!!': in the QuerySubmitted event "e.ChosenSuggestion == null" when selection an option from the drop down. This is wrong

[AutoSuggestBox.zip](https://github.com/unoplatform/uno/files/7834745/AutoSuggestBox.zip)

### Expected behavior

Should work as in WindowsAppSDK desktop

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by aesteves900**
GitHub Issue (If applicable): closes #7778

## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the current behavior?
Please see attached sample - which is a modified version of what I posted at https://github.com/unoplatform/uno/issues/4382 - and search for '!!!': in the QuerySubmitted event "e.ChosenSuggestion == null" when selection an option from the drop down. This is wrong

## What is the new behavior?
Should work as in WindowsAppSDK desktop

## PR Checklist
Please check if your PR fulfills the following requirements:
- [x] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    x:Class="UITests.Shared.Windows_UI_Xaml_Controls.AutoSuggestBoxTests.AutoSuggestBoxChosenSuggestion"
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"   
+    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+    mc:Ignorable="d"    
+    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
+
+    <Grid>
+      <TextBlock Text="Hello, world!" Margin="20" FontSize="30" />
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml:11]: Yes, just a non-used tag. I will remove it.

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+		private Book book = new Book { Author = new Author { Name = "A0" } };
+
+        public AutoSuggestBoxChosenSuggestion()
+        {
+            this.InitializeComponent();
+
+			AutoSuggestBox	autoSuggestBox;
+			Content = new Grid
+			{
+				RowDefinitions =
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs]: @jeromelaban, the UI test is not possible because the keyboard function is not allowed, yet. I did a runtime test, take a look, please.

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        {
+            if (row != 0)
+				element.SetValue(Grid.RowProperty, row);
+
+            if (span != 1)
+				element.SetValue(Grid.RowSpanProperty, span);
+
+            return element;
+        }
+	}
```
Comment (jeromelaban) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs:105]: You should not need this block. Use existing techniques to write this, and move to the XAML as needed.

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+					Content = "e.ChosenSuggestion == NULL: this is wrong !!!",
+					PrimaryButtonText = "OK",
+					RequestedTheme = ElementTheme.Default,
+					XamlRoot = XamlRoot,
+				}.ShowAsync();
+			}
+			else
+			{
+				logger.Text = $"ChosenSuggestion -> {e.ChosenSuggestion?.ToString()}";
+			}
```
Comment (Xiaoy312) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs:96]: should still update the logger text in either case:
```suggestion
			logger.Text = $"ChosenSuggestion -> {(e.ChosenSuggestion?.ToString() ?? "<null>")}";
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs [hunk 5]
```diff
@@ -96,5 +96,50 @@ static void GettingFocus(object sender, GettingFocusEventArgs e)
 				await WindowHelper.WaitForIdle();
 			}
 		}
+
+		[TestMethod]
+		public async Task When_Choose_Selection()
+		{
+			var SUT = new AutoSuggestBox();
+
+			static void QuerySubmitted(AutoSuggestBox sender, AutoSuggestBoxQuerySubmittedEventArgs args)
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs:105]: Do we have time to refactor the user example or not? The main problem was fixed.
@jeromelaban 
@francoistanguay

File: src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs [hunk 6]
```diff
@@ -325,8 +326,13 @@ private void OnQueryButtonClick(object sender, RoutedEventArgs e)
 		}
 
 		private void SubmitSearch()
-		{
-			QuerySubmitted?.Invoke(this, new AutoSuggestBoxQuerySubmittedEventArgs(_suggestionsList.SelectedItem, Text));
+	    {
+			if (_suggestionsList.Items.FirstOrDefault(it => it.ToString() == Text) is not null)
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs:330]: It only displays the results.

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+														new Author { Name = "B2" },
+														new Author { Name = "B3" },
+														new Author { Name = "a0" },
+														new Author { Name = "a1" },
+														new Author { Name = "a2" },
+														new Author { Name = "a3" },
+											};
+
+		public string Name
+		{ get; set; } = string.Empty;
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs:33]: Done!

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs (hunk 4 above)
Comment (aesteves900) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs:96]: Done!

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+				logger.Text = $"ChosenSuggestion -> {e.ChosenSuggestion?.ToString()}";
+			}
+		}
+
+		private void autoSuggestBox_TextChanged(AutoSuggestBox s, AutoSuggestBoxTextChangedEventArgs e)
+		{
+			if (e.Reason == AutoSuggestionBoxTextChangeReason.UserInput)
+			{
+				if (!string.IsNullOrEmpty(s.Text?.Trim()))
+					s.ItemsSource = Author.All.Where(a => a.Name.StartsWith(s.Text?.Trim())).ToList();
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/AutoSuggestBoxTests/AutoSuggestBoxChosenSuggestion.xaml.cs:104]: Done!

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs:105]: This condition "args.ChosenSuggestion is null" is about the bug. It is not null the component works, otherwise not.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs [hunk 9]
```diff
@@ -96,5 +96,50 @@ static void GettingFocus(object sender, GettingFocusEventArgs e)
...
+				await WindowHelper.WaitForIdle();
+				
+
+				SUT.Focus(FocusState.Programmatic);				
+				SUT.Text = "ab";
+				await WindowHelper.WaitForIdle();
+			}
+			finally
+			{
+				SUT.QuerySubmitted += QuerySubmitted;
```
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs:139]: Removed!

**Comment by MartinZikmund** [comment]
@aesteves900 This commit is missing space after `fix:` which makes it fail validation:
image

Please reword it to fix it 👍

**Review by MartinZikmund — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs]: The method was refactored. thank you!

File: src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs [hunk 10]
```diff
@@ -325,8 +326,11 @@ private void OnQueryButtonClick(object sender, RoutedEventArgs e)
 		}
 
 		private void SubmitSearch()
-		{
-			QuerySubmitted?.Invoke(this, new AutoSuggestBoxQuerySubmittedEventArgs(_suggestionsList.SelectedItem, Text));
+	    {	
+			string finalResult = GetObjectText(Text);
+
+			QuerySubmitted?.Invoke(this, new AutoSuggestBoxQuerySubmittedEventArgs(finalResult is "" ? null : finalResult, userInput));
```
Comment (kazo0) [inline src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs:332]: I'm sorry but I'm going to be annoying and ask for more clarification :) It is very possible that I am just not understanding but wouldn't this not solve the underlying issue?

According to the [AutoSuggestBox official docs](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.autosuggestbox?view=windows-app-sdk-1.1#query-submitted) the `ChosenSuggestion` object should only be set if an item was actually selected from the list, otherwise, if I type "aaa" and hit Enter, the QuerySubmitted EventArgs should have a null `ChosenSuggestion` and a `QueryText` of "aaa". 

> The QuerySubmitted event occurs when a user commits a query string. The user can commit a query in one of these ways:
> - While focus is in the text box, press Enter or click the query icon. The event args [ChosenSuggestion](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.autosuggestboxquerysubmittedeventargs.chosensuggestion?view=windows-app-sdk-1.1#microsoft-ui-xaml-controls-autosuggestboxquerysubmittedeventargs-chosensuggestion) property is null.
> - While focus is in the suggestion list, press Enter, click, or tap an item. The event args [ChosenSuggestion](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.autosuggestboxquerysubmittedeventargs.chosensuggestion?view=windows-app-sdk-1.1#microsoft-ui-xaml-controls-autosuggestboxquerysubmittedeventargs-chosensuggestion) property contains the item that was selected from the list. In all cases, the event args [QueryText](https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.autosuggestboxquerysubmittedeventargs.querytext?view=windows-app-sdk-1.1#microsoft-ui-xaml-controls-autosuggestboxquerysubmittedeventargs-querytext) property contains the text from the text box.

Looking at this code, it seems we are always setting the `ChosenSuggestion` to be the query string.

Also, I don't think we can always assume that `ChosenSuggestion` is a string. The `AutoSuggestBox` is an `ItemsControl` so the `ChosenSuggestion` should be the same type as the `ItemsSource`.

File: src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs [hunk 11]
```diff
@@ -325,8 +326,11 @@ private void OnQueryButtonClick(object sender, RoutedEventArgs e)
 		}
 
 		private void SubmitSearch()
-		{
-			QuerySubmitted?.Invoke(this, new AutoSuggestBoxQuerySubmittedEventArgs(_suggestionsList.SelectedItem, Text));
+	    {	
+			string finalResult = GetObjectText(Text);
```
Comment (kazo0) [inline src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs:330]: I will create a new issue explaining the situation with SelectedItem for AutoSuggestBox and these PR changes can be the workaround for now I guess. Perhaps for now we should at least try to set it to SelectedItem and use the Text as a fallback? Something like: var finalResult = _suggestionsList.SelectedItem ?? GetObjectText(Text);?

**Review by kazo0 — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs (hunk 11 above)
Comment (kazo0) [inline src/Uno.UI/UI/Xaml/Controls/AutoSuggestBox/AutoSuggestBox.cs:330]: I will create a new issue explaining the situation with SelectedItem for AutoSuggestBox and these PR changes can be the workaround for now i guess. Perhaps for now we should at least try to set it to SelectedItem and use the Text as a fallback? Something like: 

```suggestion
var finalResult = _suggestionsList.SelectedItem ?? GetObjectText(Text);
```
